src/ripple/app/tx/impl/applySteps.cpp

```diff
@@ -39,56 +38,122 @@
 
 namespace ripple {
 
+// Templates so preflight does the right thing with T::ConsequencesFactory.
+//
+//For Transactor::Normal
+template <class T,
+    std::enable_if_t<T::ConsequencesFactory == Transactor::Normal, int> = 0>
+std::pair<NotTEC, TxConsequences>
+invoke_preflight_helper(PreflightContext const& ctx)
+{
+    return {T::preflight(ctx), TxConsequences{ctx.tx}};
+};
+
+//For Transactor::Blocker
+template <class T,
+    std::enable_if_t<T::ConsequencesFactory == Transactor::Blocker, int> = 0>
+std::pair<NotTEC, TxConsequences>
+invoke_preflight_helper(PreflightContext const& ctx)
+{
+    return {T::preflight(ctx), TxConsequences{ctx.tx, TxConsequences::blocker}};
+};
+
+// For Transactor::Custom
+template <class T,
+    std::enable_if_t<T::ConsequencesFactory == Transactor::Custom, int> = 0>
+std::pair<NotTEC, TxConsequences>
+invoke_preflight_helper(PreflightContext const& ctx)
+{
+    return {T::preflight(ctx), T::makeTxConsequences(ctx)};
+};
+
```

Comment:
I see that this was changed to using `enable_if` instead of `if constexpr` (from the origional patch). I think I prefer using `if constexpr` in a single function over three overloads and `enable_if`, just for code clarity. 

I _suspect_ that the reason is I put an `assert` instead of an `static_assert` for if a new `ConsequencesFactory` is added, and the `enable_if` implementation will give a compile error in this case. If that's true, I think I can modify the `if constexpr` solution to use `static_assert` by calling `static_assert(sizeof(T) == -1, ""). I.e. something like:
```
template <class T>
std::pair<NotTEC, TxConsequences>
invoke_preflight_helper(PreflightContext const& ctx)
{
    TxConsequences const conseq = [&]() {
        if constexpr (T::ConsequencesFactory == Transactor::Normal)
        {
            return TxConsequences{ctx.tx};
        }
        else if constexpr (T::ConsequencesFactory == Transactor::Blocker)
        {
            return TxConsequences{ctx.tx, TxConsequences::blocker};
        }
        else if constexpr (T::ConsequencesFactory == Transactor::Custom)
        {
            return T::MakeTxConsequences(ctx);
        }
        else
        {
            static_assert(sizeof(T) == -1, "Unknown consequences factory type");
        }
    }();
    return {T::preflight(ctx), conseq};
}
```
What do you think?

Reply:
Actually I really like (and prefer) the `if constexpr` approach you coded up.  But when I ran that approach through CI, the Visual Studio complier choked on it.  GCC and Clang were okay.  And the latest Visual Studio (2019?) was okay with it.  But the older current Visual Studio couldn't handle it.

So I patched the code up using the `enable_if` approach.  It's harder for humans to read, but all of the compilers could chew through it.  It's a shame.  The `if constexpr` approach is way easier to read.

Reply:
I don't think we can justify upping visual studio support to 2019 just for this change, but yeah, I agree it's a shame we have work around this limitation.